### PR TITLE
Cl/evalutate user

### PR DESCRIPTION
**Pull request by LozanoCyrielle**
To Evaluate a loan we need :
- A pop-up for a user to evaluate a loan
- An history of all loans you were part of
- A view model that takes this rating, computes the new rating, and the new comment: put those in the DB

So I created :

- the evaluation pop-up along with its tests : Rate and comment only once for each loan
- viewModel for the evaluation pop up (and tests)
- the history of all finished loan screen (and tests) : display all the finished loans, and permit to see info or rate the loan
- viewModel for the history screen (and tests)

I also integrate the old loans screens by adding a button in the inventory screen "see old" and fix some tests

### CODE REVIEW

File: app/src/main/java/com/android/partagix/ui/components/EvaluationPopUp.kt [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+import com.android.partagix.model.loan.Loan
+
+/**
+ * EvaluationPopUp is a composable function that displays a dialog to rate and comment on a loan.
+ *
+ * @param modifier Modifier to be applied to the layout.
+ * @param loan Loan to be evaluated.
+ * @param userId Id of the user evaluating the loan.
+ * @param viewModel EvaluationViewModel to handle the evaluation.
+ */
```
Comment (skyrag) [inline app/src/main/java/com/android/partagix/ui/components/EvaluationPopUp.kt:52]: lack the onclose parameter on the javadoc

File: app/src/main/java/com/android/partagix/model/FinishedLoansViewModel.kt [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+import com.android.partagix.model.category.Category
+import com.android.partagix.model.item.Item
+import com.android.partagix.model.loan.Loan
+import com.android.partagix.model.loan.LoanState
+import com.android.partagix.model.visibility.Visibility
+import java.util.concurrent.CountDownLatch
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.StateFlow
+
+class FinishedLoansViewModel(db: Database = Database(), latch: CountDownLatch = CountDownLatch(1)) :
```
Comment (skyrag) [inline app/src/main/java/com/android/partagix/model/FinishedLoansViewModel.kt:15]: the latch isnt use in the tests if nobody is waiting for the latchcountdown you can delete the latch and the latch countdown

**Review by skyrag — CHANGES_REQUESTED** [review]
great code, some changes to be done, and some things are confusing:
- the 2 button are confusing when doing a review we don't know if one button will finish all the review or if it will only register the rate or the comment maybe do only one button
- and in the see old we can see our old loans however in the case when we are the lender we can't see to whom we lended, and if we are the borrower we can only see the owner if we check the info of the item other, and we only have the beginning of the loan and not the end of the loan

**Review by AntoineBut — CHANGES_REQUESTED** [review]
Sometimes, I was able to leave my review multiple times, I am not sure why this happened and was not able to find a reliable way to reproduce the bug.

File: app/src/main/java/com/android/partagix/ui/screens/OldLoansScreen.kt
[the remark this answers is not in the document]
Comment (LozanoCyrielle) [reply app/src/main/java/com/android/partagix/ui/screens/OldLoansScreen.kt:1]: When we have multiple old loans it was opening the pop up for all the loans, it's fixed

File: app/src/androidTest/java/com/android/partagix/screens/EvaluationPopUpScreen.kt [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+    ComposeScreen<EvaluationPopUp>(
+        semanticsProvider = semanticsProvider,
+        viewBuilderAction = { hasTestTag("evaluationPopUp") }) {
+  val rateText: KNode = onNode { hasTestTag("rateText") }
+  val rateStars: KNode = onNode { hasTestTag("rateStars") }
+  val validateButton: KNode = onNode { hasTestTag("validateButton") }
+  val commentText: KNode = onNode { hasTestTag("commentText") }
+  val commentField: KNode = onNode { hasTestTag("commentField") }
+  val commentButton: KNode = onNode { hasTestTag("commentButton") }
+  val closeButton: KNode = onNode { hasTestTag("closeButton") }
```
[the remark this answers is not in the document]
Comment (LozanoCyrielle) [reply app/src/androidTest/java/com/android/partagix/screens/EvaluationPopUpScreen.kt:17]: I changed the UI, si now there is only one button

File: app/src/androidTest/java/com/android/partagix/EvaluationViewModelTests.kt [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+  val loan2 =
+      Loan(
+          "",
+          "idOwner2",
+          "idLoaner2",
+          "item2",
+          Date(),
+          Date(),
+          "2.0",
+          "2.0",
```
[the remark this answers is not in the document]
Comment (LozanoCyrielle) [reply app/src/androidTest/java/com/android/partagix/EvaluationViewModelTests.kt:45]: There is one, where you didn't leave a rating but I can put a loan with no rating at all

File: app/src/main/java/com/android/partagix/model/FinishedLoansViewModel.kt (hunk 2 above)
Comment (LozanoCyrielle) [reply app/src/main/java/com/android/partagix/model/FinishedLoansViewModel.kt:15]: Nice catch thanks !

File: app/src/main/java/com/android/partagix/ui/components/EvaluationPopUp.kt [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+                    modifier
+                        .fillMaxWidth()
+                        .padding(start = 16.dp, end = 0.dp, bottom = 16.dp, top = 0.dp)
+                        .background(MaterialTheme.colorScheme.background)
+                        .testTag("evaluationPopUp")) {
+                  Row(
+                      horizontalArrangement = Arrangement.SpaceBetween,
+                      verticalAlignment = Alignment.CenterVertically,
+                      modifier = modifier.fillMaxWidth()
+                      // .padding(start= 10.dp, end = 26.dp)
```
[the remark this answers is not in the document]
Comment (LozanoCyrielle) [reply app/src/main/java/com/android/partagix/ui/components/EvaluationPopUp.kt]: Yes ! Done !

File: app/src/main/java/com/android/partagix/ui/components/EvaluationPopUp.kt (hunk 1 above)
Comment (LozanoCyrielle) [reply app/src/main/java/com/android/partagix/ui/components/EvaluationPopUp.kt:52]: I forgot to update it, I fixed that !

File: app/src/main/java/com/android/partagix/ui/screens/OldLoansScreen.kt [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+              IconButton(
+                  modifier = modifier.fillMaxWidth(),
+                  colors =
+                      IconButtonColors(
+                          containerColor = MaterialTheme.colorScheme.primary,
+                          contentColor = MaterialTheme.colorScheme.onPrimary,
+                          disabledContentColor = MaterialTheme.colorScheme.onPrimary,
+                          disabledContainerColor = MaterialTheme.colorScheme.primary),
+                  onClick = {
+                    println("---------------------------------cliclic")
```
[the remark this answers is not in the document]
Comment (LozanoCyrielle) [reply app/src/main/java/com/android/partagix/ui/screens/OldLoansScreen.kt]: Deleted !

File: app/src/main/java/com/android/partagix/ui/screens/OldLoansScreen.kt [hunk 7]
```diff
@@ -0,0 +1,231 @@
...
+}
+
+/**
+ * ExpandableCard is a composable function that displays a card that can be expanded to show more
+ * information.
+ *
+ * @param modifier Modifier to be applied to the layout.
+ * @param loan Loan to be displayed.
+ * @param item Item to be displayed.
+ */
```
[the remark this answers is not in the document]
Comment (LozanoCyrielle) [reply app/src/main/java/com/android/partagix/ui/screens/OldLoansScreen.kt:152]: I forgot to update it, I fixed that !

File: app/src/main/java/com/android/partagix/model/EvaluationViewModel.kt [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+
+import androidx.lifecycle.ViewModel
+import com.android.partagix.model.loan.Loan
+import com.android.partagix.model.loan.LoanState
+import java.util.Date
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.StateFlow
+
+class EvaluationViewModel(
+    loan: Loan = Loan("", "", "", "", Date(), Date(), "", "", "", "", LoanState.FINISHED),
```
[the remark this answers is not in the document]
Comment (LozanoCyrielle) [reply app/src/main/java/com/android/partagix/model/EvaluationViewModel.kt:11]: There is already an empty loan from the last merge, so I use it now, nice catch !

File: app/src/main/java/com/android/partagix/model/FinishedLoansViewModel.kt [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+      _uiState.value = _uiState.value.copy(loans = emptyList())
+      // TODO: Handle error
+      latch.countDown()
+      return
+    } else {
+      database.getLoans {
+        _uiState.value = _uiState.value.copy(loans = emptyList())
+        it.filter { loan ->
+              loan.state == LoanState.FINISHED &&
+                  (loan.idLender == user.uid || loan.idBorrower == user.uid)
```
[the remark this answers is not in the document]
Comment (LozanoCyrielle) [reply app/src/main/java/com/android/partagix/model/FinishedLoansViewModel.kt:51]: I know it is, but this code is readable and people can comprehend it. I think we leave it like this as it's not something dark for someone reading the code

File: app/src/main/java/com/android/partagix/ui/components/EvaluationPopUp.kt [hunk 10]
```diff
@@ -0,0 +1,244 @@
...
+  }
+
+  if (openDialog.value) {
+    Dialog(
+        onDismissRequest = {
+          if (haveRated.value) {
+            viewModel.reviewLoan(loan, rating, comment, idReviewed)
+          }
+          openDialog.value = false
+          onClose()
```
[the remark this answers is not in the document]
Comment (LozanoCyrielle) [reply app/src/main/java/com/android/partagix/ui/components/EvaluationPopUp.kt]: I refactored the evaluation pop as it was perfectly working, I think this is not a problem anymore

File: app/src/main/java/com/android/partagix/ui/components/ItemListSeparator.kt [hunk 11]
```diff
@@ -67,7 +70,12 @@ fun ItemListColumn(
               ),
           modifier = Modifier.fillMaxWidth(0.5f).padding(horizontal = 10.dp))
       if (canSeeOld) {
-        Text(text = "see Old", modifier = Modifier.fillMaxWidth(0.5f).clickable { /* TO-DO */})
+        Text(
```
[the remark this answers is not in the document]
Comment (LozanoCyrielle) [reply app/src/main/java/com/android/partagix/ui/components/ItemListSeparator.kt:73]: For the clickable button I will assume you talk about the one in the inventory and not this one: for now with not outgoing request we don't see this component.
As for the clarity of this button : we already have several clickable text like that one so I think I should leave it like this for now, and when we decide that those types of clickable text are not clear, we will change them all. But for the sake of uniformity in the app for now it's best its stay this way.

File: app/src/androidTest/java/com/android/partagix/components/EvaluationPopUpTest.kt [hunk 12]
```diff
@@ -0,0 +1,165 @@
...
+import org.junit.Rule
+import org.junit.Test
+import org.junit.runner.RunWith
+
+@RunWith(AndroidJUnit4::class)
+class EvaluationPopUpTest {
+  @get:Rule val composeTestRule = createComposeRule()
+  lateinit var evaluationViewModel: EvaluationViewModel
+  lateinit var db: Database
+  val loan1 =
```
[the remark this answers is not in the document]
Comment (LozanoCyrielle) [reply app/src/androidTest/java/com/android/partagix/components/EvaluationPopUpTest.kt]: I refactor the tests, so the name are more clear

File: app/src/main/java/com/android/partagix/model/Database.kt [hunk 13]
```diff
@@ -478,28 +478,29 @@ class Database(database: FirebaseFirestore = Firebase.firestore) {
...
+      reviewedUserId: String,
       rank: Double,
       comment: String = "",
   ) {
     getLoans { loans ->
       for (loan in loans) {
 
         if (loan.state == LoanState.FINISHED // only finished loans
-        && loan.id == loanId && loan.idLender == userId) {
+        && loan.id == loanId && loan.idLender == reviewedUserId) {
```
Comment (AntoineBut) [inline app/src/main/java/com/android/partagix/model/Database.kt:495]: This is not great for code readability and complexity, there might be a more elegant way to do it.

File: app/src/main/java/com/android/partagix/model/Database.kt [hunk 14]
```diff
@@ -478,28 +478,29 @@ class Database(database: FirebaseFirestore = Firebase.firestore) {
...
 
           this.loan.document(loanId).update("review_lender", rank.toString())
+          newAverageRank(reviewedUserId)
           if (comment != "") this.loan.document(loanId).update("comment_lender", comment)
         } else if (loan.state == LoanState.FINISHED // only finished loans
-        && loan.id == loanId && loan.idBorrower == userId) {
-
+        && loan.id == loanId && loan.idBorrower == reviewedUserId) {
```
Comment (AntoineBut) [inline app/src/main/java/com/android/partagix/model/Database.kt:501]: same here

File: app/src/androidTest/java/com/android/partagix/FinishedLoansViewModelTests.kt [hunk 15]
```diff
@@ -0,0 +1,153 @@
...
+  fun getItemTests() {
+    every { Authentication.getUser() } returns mockUser
+    every { mockUser.uid } returns "Cedric"
+    every { db.getItem("item1", any()) } answers
+        { invocation ->
+          val onSuccessItem = invocation.invocation.args[1] as (Item) -> Unit
+          onSuccessItem(item)
+        }
+    finishedLoansViewModel.getItem("item1")
+    assert(finishedLoansViewModel.uiItem.value == item)
```
[the remark this answers is not in the document]
Comment (LozanoCyrielle) [reply app/src/androidTest/java/com/android/partagix/FinishedLoansViewModelTests.kt:151]: I don't know how I can use "more descriptive assertion messages" but to make it clearer I added comments. I hope this is enough :))

File: app/src/androidTest/java/com/android/partagix/screens/EvaluationPopUpScreen.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (AntoineBut) [reply app/src/androidTest/java/com/android/partagix/screens/EvaluationPopUpScreen.kt:15]: This is very nice!

File: app/src/main/java/com/android/partagix/model/FinishedLoansViewModel.kt (hunk 9 above)
[the remark this answers is not in the document]
Comment (AntoineBut) [reply app/src/main/java/com/android/partagix/model/FinishedLoansViewModel.kt:49]: ok

File: app/src/main/java/com/android/partagix/ui/screens/OldLoansScreen.kt
[the remark this answers is not in the document]
Comment (AntoineBut) [reply app/src/main/java/com/android/partagix/ui/screens/OldLoansScreen.kt:1]: perfect, it looks good now

**Review by AntoineBut — APPROVED** [review]
Nice work, the user flow is very clear and the pop up looks very nice

File: app/src/androidTest/java/com/android/partagix/screens/EvaluationPopUpScreen.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (skyrag) [reply app/src/androidTest/java/com/android/partagix/screens/EvaluationPopUpScreen.kt:15]: nice change it is less confusing that way

**Review by skyrag — APPROVED** [review]
the ui for the button is no more confusing and we can now see both of the date and the name of the other person of the loan, good changes

**Pull request merged**